src/TypeScript/TsConfig/TsConfigsLocator.cs

```diff
@@ -21,47 +21,109 @@
+        internal const int VSPROPID_IsInOpenFolderMode = -8044;
 
-        private readonly IVsUIShellOpenDocument vsUiShellOpenDocument;
+        private readonly IVsHierarchyLocator vsHierarchyLocator;
+        private readonly IVsSolution vsSolution;
+        private readonly IFileSystem fileSystem;
+        private readonly ILogger logger;
 
         [ImportingConstructor]
-        public TsConfigsLocator([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
+        public TsConfigsLocator([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IVsHierarchyLocator vsHierarchyLocator,
+            ILogger logger)
+            : this(serviceProvider, vsHierarchyLocator, new FileSystem(), logger)
+        {
+        }
+
+        internal TsConfigsLocator(IServiceProvider serviceProvider,
+            IVsHierarchyLocator vsHierarchyLocator,
+            IFileSystem fileSystem, 
+            ILogger logger)
         {
-            vsUiShellOpenDocument = serviceProvider.GetService(typeof(SVsUIShellOpenDocument)) as IVsUIShellOpenDocument;
+            vsSolution = serviceProvider.GetService(typeof(SVsSolution)) as IVsSolution;
+            this.vsHierarchyLocator = vsHierarchyLocator;
+            this.fileSystem = fileSystem;
+            this.logger = logger;
+        }
+
+        public IReadOnlyList<string> Locate(string sourceFilePath)
+        {
+            var projectDirectory = GetProjectDirectory(sourceFilePath);
+
+            if (projectDirectory == null)
+            {
+                return Array.Empty<string>();
+            }
+
+            var foundTsConfigs = fileSystem
+                .Directory
+                .EnumerateFiles(projectDirectory, "tsconfig.json", SearchOption.AllDirectories)
+                .Where(x=> !x.Contains("\\node_modules\\"));
+
+            return foundTsConfigs.ToArray();
         }
 
-        public IReadOnlyList<string> Locate()
+        private string GetProjectDirectory(string sourceFilePath)
         {
-            var foundFiles = new string[MaxNumberOfFiles];
+            var hr = vsSolution.GetProperty(VSPROPID_IsInOpenFolderMode, out var isOpenAsFolder);
+
+            Debug.Assert(hr == VSConstants.S_OK || hr == VSConstants.E_NOTIMPL, "Failed to retrieve VSPROPID_IsInOpenFolderMode");
+
+            if (isOpenAsFolder != null && (bool) isOpenAsFolder)
+            {
+                hr = vsSolution.GetProperty((int) __VSPROPID.VSPROPID_SolutionDirectory, out var solutionDirectory);
+
+                Debug.Assert(hr == VSConstants.S_OK || hr == VSConstants.E_NOTIMPL,
+                    "Failed to retrieve VSPROPID_SolutionDirectory");
+
+                return (string) solutionDirectory;
+            }
+
+            var hierarchyToSearch = vsHierarchyLocator.GetVsHierarchyForFile(sourceFilePath);
+
+            if (hierarchyToSearch == null)
+            {
+                logger.WriteLine(Resources.ERR_NoVsHierarchy, sourceFilePath);
```

Comment:
The wording of the error message hasn't changed in this PR, but I think it would be better to say "project" than "IVSHierarchy" in the message. Almost no-one will know what an IVSHierarchy is.